src/libraries/System.IO.MemoryMappedFiles/src/System/IO/MemoryMappedFiles/MemoryMappedFile.Unix.cs

```diff
@@ -190,10 +190,20 @@ private static SafeFileHandle CreateSharedBackingObject(Interop.Sys.MemoryMapped
             do
             {
                 mapName = GenerateMapName();
-                fd = Interop.Sys.ShmOpen(mapName, flags, (int)perms); // Create the shared memory object.
+                if (Interop.Sys.IsMemfdSupported)
+                {
+                    fd = Interop.Sys.MemfdCreate(mapName);
+                    // none of the errors are transient/recoverable at this point
+                    if (fd.IsInvalid) throw Interop.GetExceptionForIoErrno(Interop.Sys.GetLastErrorInfo());
```

Comment:
Why is this needed rather than just relying on the shared fd.IsInvalid check after the if/else?

Reply:
We had a discussion that none of the `memfd_create()` error cases are transient in nature: https://github.com/dotnet/runtime/pull/105178#discussion_r1686175008, so @adamsitnik suggested that we early bail from loop (or skip the loop altogether).

Reply:
> We had a discussion that none of the memfd_create() error cases are transient in nature: https://github.com/dotnet/runtime/pull/105178#discussion_r1686175008

From that list, all of these are technically transient in nature:
```
       EMFILE The per-process limit on the number of open file
              descriptors has been reached.

       ENFILE The system-wide limit on the total number of open files
              has been reached.

       ENOMEM There was insufficient memory to create a new anonymous
              file.
```

The loop is there to handle EEXIST and it will only loop if the error is EEXIST. Since that's not an error that'll ever come from memfd_create, the loop is a nop for that.

Either we should use the same shared error handling after the if/else for both conditions, or that subsequent error handling block should be moved up into the else, as it's then specific to the shmopen case.

Reply:
Changed to use shared error handling.